addons/mail/static/src/messaging/entity/entity_core.js

```diff
@@ -0,0 +1,246 @@
+                    }/${
+                        relation.inverse
+                    }" misconfigured (currently "${
+                        inverseRelation.to
+                    }", should instead refer to this entity or parented entity: ${
+                        allSelfAndParentNames.map(name => `"${name}"`).join(', ')
+                    }?)`
+                );
+            }
+            if (
+                (relation.type === 'many2many' && inverseRelation.type !== 'many2many') ||
+                (relation.type === 'one2one' && inverseRelation.type !== 'one2one') ||
+                (relation.type === 'one2many' && inverseRelation.type !== 'many2one') ||
+                (relation.type === 'many2one' && inverseRelation.type !== 'one2many')
+            ) {
+                throw new Error(
+                    `Mismatch relations types "${
+                        Entity.name
+                    }/${
+                        relationName
+                    }" (${
+                        relation.type
+                    }) and "${
+                        RelatedEntity.name
+                    }/${
+                        relation.inverse
+                    }" (${
+                        inverseRelation.type
+                    }).`
+                );
+            }
+        }
+    }
+}
+
+/**
+ * @returns {Object}
+ */
+function generateEntities() {
+    const allNames = Object.keys(registry);
+    const Entities = {};
+    const generatedNames = [];
+    let toGenerateNames = [...allNames];
+    while (toGenerateNames.length > 0) {
+        const generateable = toGenerateNames.map(name => registry[name]).find(entry => {
+            let isGenerateable = true;
+            for (const dependencyName of entry.dependencies) {
+                if (!generatedNames.includes(dependencyName)) {
+                    isGenerateable = false;
+                }
+            }
+            return isGenerateable;
+        });
+        if (!generateable) {
+            throw new Error(`Cannot generate following Entity classes: ${toGenerateNames.split(', ')}`);
+        }
+        const Entity = generateable.factory(Entities);
+        for (const patch of generateable.patches) {
+            if (patch.isOnClass) {
+                patchClassMethods(Entity, patch.name, patch.patch);
+            } else {
+                patchInstanceMethods(Entity, patch.name, patch.patch);
+            }
+        }
+        Entities[Entity.name] = Entity;
+        generatedNames.push(Entity.name);
+        toGenerateNames = toGenerateNames.filter(name => name !== Entity.name);
+    }
+    return Entities;
+}
+
+/**
+ *
+ * @param {string} entityName
+ * @param {string} patchName
+ * @param {Object} patch
+ */
+function registerClassPatchEntity(entityName, patchName, patch) {
+    _registerPatchEntity(entityName, patchName, patch, { isOnClass: true });
+}
+
+/**
+ * FIXME: instance patch are not supported due to patch making changes on
```

Comment:
fixme remaining

Reply:
kinda intended to keep it like this, because we likely want to patch entity instance methods but we can't due to how both patch and observable work... Workaround is to patch class methods, therefore instance method should call class methods in order to be patchable